zerver/models.py

```diff
@@ -750,38 +750,63 @@ def flush_realm_emoji(sender: Any, **kwargs: Any) -> None:
 post_save.connect(flush_realm_emoji, sender=RealmEmoji)
 post_delete.connect(flush_realm_emoji, sender=RealmEmoji)
 
-def filter_pattern_validator(value: str) -> None:
-    regex = re.compile(r'^(?:(?:[\w\-#_= /:]*|[+]|[!])(\(\?P<\w+>.+\)))+$')
-    error_msg = _('Invalid filter pattern.  Valid characters are {}.').format(
-        '[ a-zA-Z_#=/:+!-]',)
-
-    if not regex.match(str(value)):
-        raise ValidationError(error_msg)
-
-    try:
-        re.compile(value)
-    except re.error:
-        # Regex is invalid
-        raise ValidationError(error_msg)
-
-def filter_format_validator(value: str) -> None:
-    regex = re.compile(r'^([\.\/:a-zA-Z0-9#_?=&;~-]+%\(([a-zA-Z0-9_-]+)\)s)+[/a-zA-Z0-9#_?=&;~-]*$')
-
-    if not regex.match(value):
-        raise ValidationError(_('Invalid URL format string.'))
-
 class RealmFilter(models.Model):
     """Realm-specific regular expressions to automatically linkify certain
     strings inside the Markdown processor.  See "Custom filters" in the settings UI.
     """
     id: int = models.AutoField(auto_created=True, primary_key=True, verbose_name='ID')
     realm: Realm = models.ForeignKey(Realm, on_delete=CASCADE)
-    pattern: str = models.TextField(validators=[filter_pattern_validator])
-    url_format_string: str = models.TextField(validators=[URLValidator(), filter_format_validator])
+    pattern: str = models.TextField()
+    url_format_string: str = models.TextField()
 
     class Meta:
         unique_together = ("realm", "pattern")
 
+    def clean(self) -> None:
+        # Django's `full_clean` calls `clean_fields` followed by `clean` and
+        # stores all ValidationErrors from all stages to return as JSON. Thus,
+        # these validations are in addition to field specific validations.
+
+        # extract substitutions
+        try:
+            pattern = re.compile(self.pattern)
+        except re.error:
+            # Regex is invalid
+            raise ValidationError(
+                _('Invalid filter pattern. Please check the pattern for syntax errors.')
+            )
+
+        group_set = set(pattern.groupindex.keys())
+        group_match_regex = re.compile(r'%\((?P<group_name>[^)]+)\)s')
```

Comment:
This is not correct in the presence of `%%` (which is how you embed a literal `%` in a format string).

Group names aren’t allowed to include `(` either.

There’s no point in using `re.compile` on a regex you’re throwing away after one use.

Reply:
> This is not correct in the presence of `%%` (which is how you embed a literal `%` in a format string).

Can you give an example so that I can understand this more better ?

Reply:
```pycon
>>> "Hello %(foo)s" % {"foo": "bar"}
'Hello bar'
>>> "Hello %%(foo)s" % {"foo": "bar"}
'Hello %(foo)s'
>>> "Hello %%%(foo)s" % {"foo": "bar"}
'Hello %bar'
>>> "Hello %%%%(foo)s" % {"foo": "bar"}
'Hello %%(foo)s'
>>> "Hello %%%%%(foo)s" % {"foo": "bar"}
'Hello %%bar'
>>> "Hello %(inv(alid)s" % {"inv(alid": "bar"}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: incomplete format key
```